partiql-ast/src/main/java/org/partiql/ast/v1/AstVisitor.java

```diff
@@ -0,0 +1,367 @@
+import org.partiql.ast.v1.type.TypeNull;
+import org.partiql.ast.v1.type.TypeNumeric;
+import org.partiql.ast.v1.type.TypeReal;
+import org.partiql.ast.v1.type.TypeSexp;
+import org.partiql.ast.v1.type.TypeSmallint;
+import org.partiql.ast.v1.type.TypeString;
+import org.partiql.ast.v1.type.TypeStruct;
+import org.partiql.ast.v1.type.TypeSymbol;
+import org.partiql.ast.v1.type.TypeTime;
+import org.partiql.ast.v1.type.TypeTimeWithTz;
+import org.partiql.ast.v1.type.TypeTimestamp;
+import org.partiql.ast.v1.type.TypeTimestampWithTz;
+import org.partiql.ast.v1.type.TypeTinyint;
+import org.partiql.ast.v1.type.TypeTuple;
+import org.partiql.ast.v1.type.TypeVarchar;
+
+public interface AstVisitor<R, C> {
+    R visit(AstNode node, C ctx);
+
+    R visitStatement(Statement node, C ctx);
+
+    R visitQuery(Query node, C ctx);
+
+    R visitExplain(Explain node, C ctx);
+
+    R visitTarget(Target node, C ctx);
+
+    R visitTargetDomain(Target.Domain node, C ctx);
+
+    R visitType(Type node, C ctx);
+
+    R visitTypeNull(TypeNull node, C ctx);
+
+    R visitTypeMissing(TypeMissing node, C ctx);
+
+    R visitTypeBool(TypeBool node, C ctx);
+
+    R visitTypeTinyint(TypeTinyint node, C ctx);
+
+    R visitTypeSmallint(TypeSmallint node, C ctx);
+
+    R visitTypeInt2(TypeInt2 node, C ctx);
+
+    R visitTypeInt4(TypeInt4 node, C ctx);
+
+    R visitTypeBigint(TypeBigint node, C ctx);
+
+    R visitTypeInt8(TypeInt8 node, C ctx);
+
+    R visitTypeInt(TypeInt node, C ctx);
+
+    R visitTypeReal(TypeReal node, C ctx);
+
+    R visitTypeFloat32(TypeFloat32 node, C ctx);
+
+    R visitTypeFloat64(TypeFloat64 node, C ctx);
+
+    R visitTypeDecimal(TypeDecimal node, C ctx);
+
+    R visitTypeNumeric(TypeNumeric node, C ctx);
+
+    R visitTypeChar(TypeChar node, C ctx);
+
+    R visitTypeVarchar(TypeVarchar node, C ctx);
+
+    R visitTypeString(TypeString node, C ctx);
+
+    R visitTypeSymbol(TypeSymbol node, C ctx);
+
+    R visitTypeBit(TypeBit node, C ctx);
+
+    R visitTypeBitVarying(TypeBitVarying node, C ctx);
+
+    R visitTypeByteString(TypeByteString node, C ctx);
+
+    R visitTypeBlob(TypeBlob node, C ctx);
+
+    R visitTypeClob(TypeClob node, C ctx);
+
+    R visitTypeDate(TypeDate node, C ctx);
+
+    R visitTypeTime(TypeTime node, C ctx);
+
+    R visitTypeTimeWithTz(TypeTimeWithTz node, C ctx);
+
+    R visitTypeTimestamp(TypeTimestamp node, C ctx);
+
+    R visitTypeTimestampWithTz(TypeTimestampWithTz node, C ctx);
+
+    R visitTypeInterval(TypeInterval node, C ctx);
+
+    R visitTypeBag(TypeBag node, C ctx);
+
+    R visitTypeList(TypeList node, C ctx);
+
+    R visitTypeSexp(TypeSexp node, C ctx);
+
+    R visitTypeTuple(TypeTuple node, C ctx);
+
+    R visitTypeStruct(TypeStruct node, C ctx);
+
+    R visitTypeAny(TypeAny node, C ctx);
+
+    R visitTypeCustom(TypeCustom node, C ctx);
+
+    R visitIdentifier(Identifier node, C ctx);
+
+    R visitIdentifierSymbol(Identifier.Symbol node, C ctx);
```

Comment:
(self-review) I tried to get rid of excessive nesting of classes from the original sprout-generated AST code. Commit [`1a4e2de` (#1579)](https://github.com/partiql/partiql-lang-kotlin/pull/1579/commits/1a4e2de73ce9a6d3f329472af0d8fcf618eaf0ee) shows the change to use at most 1 nested class in the Kotlin code before translating the classes and interfaces to Java.